### PR TITLE
Refactor Battle Animations

### LINKED ISSUE #1757: Map battle animations are not frame accurate
Currently we animate battle animations at the beginning of `Game_Map::Update()`.

From the below tests, this appears wrong.

We also don't correctly save and load all of the `battleanim_` chunks.

### PR DESCRIPTION
**Pull request by fmatthew5876**
Fix: #1757 
Fix: #1372

~~Depends on #1784~~

This supports RPG_RT compatible saving and loading of map battle animations

TBD:
- [x] battle animation screen effects vs event screen effects.
    - [x] Do they use the LSD chunks and affect saving and loading?
- [x] battle animation screen effects and sounds when start battle before anim finishes
    - [x] What if a battle animation in the battle runs? Will it interrupt the map anim?
    - [x] Loading a save before battle?
    - [x] Can we make a very long battle anim with a very short battle that conyinues running after battle exits?
- [x] Frame specific behavior of timing effects when load game
- [x] Verify flash length with OpenSaveMenu test
- [x] Behavior when screen flash is active when animation starts
- [x] Behavior when other event flashes while anim is running

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This now passes all test cases in #1757 (except shake screen comments)

We also accurately emulate the weird behaviors that occur when you start a battle while a battle animation is playing on the map. #1372

**Comment by fmatthew5876** [comment]
I've fixed the parameters for animation shake screen and implemented shake target.

I think my parameters are pretty close but they aren't perfect. I've tried various screen capture methods and this is the best I've got so far. The only thing I'm pretty certain of is that they animate for 32 frames.

Would be really great to know the algo used for these. @CherryDT ?? :)

This PR is finished and ready for review

File: src/game_map.h [hunk 1]
```diff
@@ -363,10 +363,11 @@ namespace Game_Map {
 	 * @param animation_id the animation ID
 	 * @param target_id the ID of the targeted character
 	 * @param global whether to "show on the entire map"
+	 * @param which frame to start on.
```
Comment (Ghabry) [inline src/game_map.h]: >  Support loading saved games with map battle animations. 
doc comment: start_frame

File: src/game_battlealgorithm.cpp [hunk 2]
```diff
@@ -284,7 +282,7 @@ void Game_BattleAlgorithm::AlgorithmBase::PlaySoundAnimation(bool on_source, int
 
 	if (on_source) {
 		std::vector<Game_Battler*> anim_targets = { GetSource() };
-		Game_Battle::ShowBattleAnimation(GetAnimation()->ID, anim_targets, false, true, cutoff);
+		Game_Battle::ShowBattleAnimation(GetAnimation()->ID, { GetSource() }, true, cutoff);
```
Comment (Ghabry) [inline src/game_battlealgorithm.cpp]: >  Remove BattleAnimation::ShouldScreenFlash() 
unused var anim_targets

File: src/game_screen.h [hunk 3]
```diff
@@ -121,6 +121,18 @@ class Game_Screen {
 	 */
 	bool IsBattleAnimationWaiting();
 
+	/**
+	 * Animates the screen shake algorithm given the parameters
+	 *
+	 * @param strength the strength of the shake
+	 * @param speed of the shake
+	 * @param time_left how much time is left in frames
+	 * @param current shake displacement
```
Comment (Ghabry) [inline src/game_screen.h]: > Abstract out the shake algorithm 

doc comment: position

**Comment by fmatthew5876** [comment]
Rebased and fixed the review comments

**Comment by Ghabry** [comment]
Jenkins: test this please

**Pull request merged**